src/audio/google/google_rtc_audio_processing.c

```diff
@@ -543,223 +548,324 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
-	if (ret)
-		return ret;
+	if (ret == 0)
+		ret = google_rtc_audio_processing_reconfigure(mod);
 
-	return 0;
+	return ret;
+}
+
+static int trigger_handler(struct processing_module *mod, int cmd)
+{
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	/* Ignore and halt propagation if we get a trigger from the
+	 * playback pipeline: not for us.
+	 */
+	if (cd->ref_comp_buffer->walking)
+		return PPL_STATUS_PATH_STOP;
+
+	/* Note: not module_adapter_set_state().  With IPC4 those are
+	 * identical, but IPC3 has some odd-looking logic that
+	 * validates that no sources are active when receiving a
+	 * PRE_START command, which obviously breaks for our reference
+	 * stream if playback was already running when our pipeline
+	 * started
+	 */
+	return comp_set_state(mod->dev, cmd);
 }
 
 static int google_rtc_audio_processing_reset(struct processing_module *mod)
 {
 	comp_dbg(mod->dev, "google_rtc_audio_processing_reset()");
-
 	return 0;
 }
 
-static int google_rtc_audio_processing_process(struct processing_module *mod,
-					       struct input_stream_buffer *input_buffers,
-					       int num_input_buffers,
-					       struct output_stream_buffer *output_buffers,
-					       int num_output_buffers)
+/* FunctionMostlyExistsToKeepLineLengthsUnderControl */
+static inline void execute_aec(struct google_rtc_audio_processing_comp_data *cd)
 {
-	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
-	int16_t *src, *dst, *ref;
-	uint32_t num_aec_reference_frames;
-	uint32_t num_aec_reference_bytes;
-	int num_samples_remaining;
-	int num_frames_remaining;
-	int channel;
-	int frames;
-	int nmax;
-	int ret;
-	int i, j, n;
-
-	struct input_stream_buffer *ref_streamb, *mic_streamb;
-	struct output_stream_buffer *out_streamb;
-	struct audio_stream *ref_stream, *mic_stream, *out_stream;
+	/* Note that reference input and mic output share the same
+	 * buffer for efficiency
+	 */
+	GoogleRtcAudioProcessingAnalyzeRender_float32(cd->state,
+						      (const float **)cd->refout_buffers);
+	GoogleRtcAudioProcessingProcessCapture_float32(cd->state,
+						       (const float **)cd->raw_mic_buffers,
+						       cd->refout_buffers);
+	cd->buffered_frames = 0;
+}
 
-	if (cd->reconfigure) {
-		ret = google_rtc_audio_processing_reconfigure(mod);
-		if (ret)
-			return ret;
+static void mic_in_copy(struct sof_source *src, int frames, float **dst_bufs, int frame0)
+{
+	size_t chan = MIN(MIC_CHAN_MAX, source_get_channels(src));
+	size_t samples = frames * chan;
+	size_t bytes = samples * sizeof(mic_sample_t);
+	const mic_sample_t *buf, *bufstart, *bufend;
+	float *dst[MIC_CHAN_MAX];
+	int i, c, err;
+	size_t bufsz;
+
+	for (i = 0; i < chan; i++)
+		dst[i] = &dst_bufs[i][frame0];
+
+	err = source_get_data(src, bytes, (void *)&buf, (void *)&bufstart, &bufsz);
+	assert(err == 0);
+	bufend = &bufstart[bufsz];
+
+	for (i = 0; i < frames; i++) {
+		for  (c = 0; c < chan; c++) {
```

Comment:
As `chan` can be different from `source_get_channels(src)` shouldn't one take car to advance `buf` by `source_get_channels(src)` between each frames ?

Reply:
Indeed, this is wrong.  (Actually it's doubly wrong: the error here doesn't overflow the buffer because the channel count from source_get_channels() is wrong, cancelling each other out but doing the wrong thing and running at effectively 20ms instead of 10).  Will respin and have the fix up this morning